ddtrace/internal/processor/stats.py

```diff
@@ -0,0 +1,243 @@
+from collections import defaultdict
+import typing
+
+import ddtrace
+from ddtrace import config
+from ddtrace.vendor.ddsketch.ddsketch import LogCollapsingLowestDenseDDSketch
+from ddtrace.vendor.ddsketch.pb.proto import DDSketchProto
+
+from . import SpanProcessor
+from ...constants import SPAN_MEASURED_KEY
+from .._encoding import packb
+from ..agent import get_connection
+from ..compat import get_connection_response
+from ..compat import time_ns
+from ..forksafe import Lock
+from ..hostname import get_hostname
+from ..logger import get_logger
+from ..periodic import PeriodicService
+from ..writer import _human_size
+
+
+if typing.TYPE_CHECKING:
+    from typing import DefaultDict
+    from typing import Dict
+    from typing import List
+    from typing import Optional
+
+    from ddtrace import Span
+
+    from ..agent import ConnectionType
+
+
+log = get_logger(__name__)
+
+
+def _is_top_level(span):
+    # type: (Span) -> bool
+    """Return whether the span is a "top level" span."""
+    return (span._local_root is span) or (
+        span._parent is not None and span._parent.service != span.service and span.service is not None
+    )
+
+
+def _is_measured(span):
+    # type: (Span) -> bool
+    """Return whether the span is flagged to be measured or not."""
+    return span.metrics.get(SPAN_MEASURED_KEY) == 1
+
+
+"""
+To aggregate metrics for spans they need to be "uniquely" identified (as
+best as possible). This enables the compression of stat points.
+
+Aggregation can be done using primary and secondary attributes from the span
+stored in a tuple which is hashable in Python.
+"""
+SpanAggrKey = typing.Tuple[
+    str,  # name
+    str,  # service
+    str,  # resource
+    str,  # type
+    int,  # http status code
+    bool,  # synthetics request
+]
+
+
+class SpanAggrStats(object):
+    """Aggregated span statistics."""
+
+    def __init__(self):
+        self.hits = 0
+        self.top_level_hits = 0
+        self.errors = 0
+        self.duration = 0
+        # Match the relative accuracy of the sketch implementation used in the backend
+        # which is 0.775%.
+        self.ok_distribution = LogCollapsingLowestDenseDDSketch(0.00775, bin_limit=2048)
+        self.err_distribution = LogCollapsingLowestDenseDDSketch(0.00775, bin_limit=2048)
+
+
+def _span_aggr_key(span):
+    # type: (Span) -> SpanAggrKey
+    """Return a hashable key that can be used to uniquely refer to the "same"
+    span.
+    """
+    service = span.service or "unnamed_service"
```

Comment:
can we document why this is `unnamed_service`?

also, the trace agent used `unnamed-python-service` as the default, should we standardize that here?